testcases/kernel/syscalls/inotify/inotify07.c

```diff
@@ -0,0 +1,244 @@
+ * otherwise, applies only to this software file.  Patent licenses, if
+ * any, provided herein do not apply to combinations of this program with
+ * other software, or any other product whatsoever.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Started by Amir Goldstein <amir73il@gmail.com>
+ *
+ * DESCRIPTION
+ *     Check that inotify work for an overlayfs directory after copy up and
+ *     drop caches.
+ *
+ *     An inotify watch pins the directory inode in cache, but not the dentry.
+ *     The watch will not report events on the directory if overlayfs does not
+ *     obtain the pinned inode to the new allocated dentry after drop caches.
+ *
+ *     The problem has been fixed by commit:
+ *       31747eda41ef "ovl: hash directory inodes for fsnotify"
+ *
+ * ALGORITHM
+ *     Add watch on an overlayfs lower directory then chmod directory and drop
+ *     dentry and inode caches. Execute operations on directory and child and
+ *     expect events to be reported on directory watch.
+ */
+
+#include "config.h"
+
+#if defined(HAVE_SYS_INOTIFY_H)
+# include <sys/inotify.h>
+#endif
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <sys/mount.h>
+#include <limits.h>
+#include "tst_test.h"
+#include "inotify.h"
+
+#if defined(HAVE_SYS_INOTIFY_H)
+
+#define EVENT_MAX 1024
+/* size of the event structure, not counting name */
+#define EVENT_SIZE  (sizeof (struct inotify_event))
+/* reasonable guess as to size of 1024 events */
+#define EVENT_BUF_LEN        (EVENT_MAX * (EVENT_SIZE + 16))
+
+#define BUF_SIZE 256
+static int fd_notify, reap_wd;
+static int wd;
+
+struct event_t {
+	char name[BUF_SIZE];
+	unsigned int mask;
+};
+
+#define OVL_MNT "ovl"
+#define DIR_NAME "test_dir"
+#define DIR_PATH OVL_MNT"/"DIR_NAME
+#define FILE_NAME "test_file"
+#define FILE_PATH OVL_MNT"/"DIR_NAME"/"FILE_NAME
+
+static int ovl_mounted;
+
+static struct event_t event_set[EVENT_MAX];
+
+static char event_buf[EVENT_BUF_LEN];
+
+void verify_inotify(void)
+{
+	int test_cnt = 0;
+
+	/*
+	 * generate sequence of events
+	 */
+	SAFE_CHMOD(DIR_PATH, 0755);
+	event_set[test_cnt].mask = IN_ISDIR | IN_ATTRIB;
+	strcpy(event_set[test_cnt].name, "");
+	test_cnt++;
+
+	SAFE_TOUCH(FILE_PATH, 0644, NULL);
+	event_set[test_cnt].mask = IN_OPEN;
+	strcpy(event_set[test_cnt].name, FILE_NAME);
+	test_cnt++;
+	event_set[test_cnt].mask = IN_CLOSE_WRITE;
+	strcpy(event_set[test_cnt].name, FILE_NAME);
+	test_cnt++;
+	event_set[test_cnt].mask = IN_ATTRIB;
+	strcpy(event_set[test_cnt].name, FILE_NAME);
+	test_cnt++;
+
+	int len, i = 0, test_num = 0;
+	if ((len = read(fd_notify, event_buf, EVENT_BUF_LEN)) == -1) {
+		tst_brk(TBROK | TERRNO,
+			"read(%d, buf, %zu) failed",
+			fd_notify, EVENT_BUF_LEN);
+
```

Comment:
The only nit for this version is that when executed on broken kernel the failure message produced is:

BROK: read(3, buf, 32768) failed: EAGAIN/EWOULDBLOCK

Which looks much more like a broken test rather than anything else. Can we please check for the EAGAIN errno here and print a bit more user friendly TFAIL message in this case?

The same applies to the next test and apart from that the code looks good to me.